src/NuGet.Core/NuGet.Frameworks/FrameworkNameProvider.cs

```diff
@@ -179,78 +176,91 @@ public bool TryGetPlatformVersion(string versionString, [NotNullWhen(true)] out
                 }
                 return Version.TryParse(versionString, out version);
             }
-
-            return false;
         }
 
         public string GetVersionString(string framework, Version version)
         {
-            var versionString = string.Empty;
+            if (version is null || IsZero(version))
+            {
+                return string.Empty;
+            }
 
-            if (version != null
-                && (version.Major > 0
-                    || version.Minor > 0
-                    || version.Build > 0
-                    || version.Revision > 0))
+            int major = version.Major > 0 ? version.Major : 0;
+            int minor = version.Minor > 0 ? version.Minor : 0;
+            int build = version.Build > 0 ? version.Build : 0;
+            int revision = version.Revision > 0 ? version.Revision : 0;
+
+            // remove all trailing zeros beyond the minor version
+            int partCount = (minor == 0, build == 0, revision == 0) switch
             {
-                var versionParts = new Stack<int>(4);
+                (true, true, true) => 1,
+                (false, true, true) => 2,
+                (_, false, true) => 3,
+                (_, _, false) => 4
+            };
 
-                versionParts.Push(version.Major > 0 ? version.Major : 0);
-                versionParts.Push(version.Minor > 0 ? version.Minor : 0);
-                versionParts.Push(version.Build > 0 ? version.Build : 0);
-                versionParts.Push(version.Revision > 0 ? version.Revision : 0);
+            // By default require the version to have 2 digits, for legacy frameworks 1 is allowed
+            var minPartCount = SingleDigitVersionFrameworks.Contains(framework) ? 1 : 2;
```

Comment:
I think there's a further optimization to avoid this set check here. 
If part is 2 or bigger than this is irrelevant. 

Not sure it makes a true difference, but in majority of the cases, the framework version specified is at least 2 digits.

Reply:
Nice! Added.

Reply:
Nice! Added.